image/icns.ksy

```diff
@@ -0,0 +1,889 @@
+            type: u4
+            doc: The width of the icon in pixels.
+          - id: height
+            type: u4
+            doc: The height of the icon in pixels.
+        seq:
+          - id: icon
+            size: width * height / 8
+            doc: |
+              The icon bitmap data,
+              as a packed sequence of 1-bit pixels,
+              in MSB-first bit order
+              (i. e. pixels that come earlier in the file go into the higher bits).
+              0 is white and 1 is black.
+          - id: mask
+            size: width * height / 8
+            doc: |
+              The icon mask data,
+              as a packed sequence of 1-bit pixels,
+              in MSB-first bit order
+              (i. e. pixels that come earlier in the file go into the higher bits).
+              0 is transparent and 1 is solid.
+      icon_x4_data:
+        doc: |
+          The data for a 4-bit color bitmap icon.
+          These icons do not contain a mask,
+          but use the mask from the 1-bit monochrome icon of the same size from the same family.
+        params:
+          - id: width
+            type: u4
+            doc: The width of the icon in pixels.
+          - id: height
+            type: u4
+            doc: The height of the icon in pixels.
+        seq:
+          - id: icon
+            size: width * height / 2
+            doc: |
+              The icon bitmap data,
+              as a packed sequence of 4-bit pixels,
+              in MSB-first bit order
+              (i. e. pixels that come earlier in the file go into the higher bits).
+              Color values come from the Mac OS default 4-bit color palette.
+      icon_x8_data:
+        doc: |
+          The data for an 8-bit color bitmap icon.
+          These icons do not contain a mask,
+          but use the mask from the 1-bit monochrome icon of the same size from the same family.
+        params:
+          - id: width
+            type: u4
+            doc: The width of the icon in pixels.
+          - id: height
+            type: u4
+            doc: The height of the icon in pixels.
+        seq:
+          - id: icon
+            size: width * height
+            doc: |
+              The icon bitmap data,
+              as a packed sequence of 8-bit pixels.
+              Color values come from the Mac OS default 8-bit color palette.
+      icon_rgb_data:
+        doc: |
+          The data for a 24-bit RGB bitmap icon.
+          These icons do not contain a mask -
+          the mask is stored as a separate element in the same icon family.
+        params:
+          - id: width
+            type: u4
+            doc: The width of the icon in pixels.
+          - id: height
+            type: u4
+            doc: The height of the icon in pixels.
+        seq:
+          - id: data
+            type:
+              switch-on: _io.size == width * height * 3
+              cases:
+                false: icns_style_packbits
```

Comment:
IMHO, for uniformity a type having exactly the same interface as `icns_style_packbits` is needed. We also would like to have all the types containing image data having the same interface. Even if we don't have interfaces in KS currently, for dynamic duck-typed languages this fact is a bit less important.

Reply:
> for uniformity a type having exactly the same interface as `icns_style_packbits` is needed in `true` branch.

How would that be possible? `icns_style_packbits` is compressed data and the byte array case is uncompressed data - they *have* to be handled differently by code that uses the spec.

> We also would like to have all the types containing image data having the same interface.

I thought about this, and I intentionally made the different image data types have similar/compatible param/field/instance names. I didn't make them all completely identical though - for example the Classic-style bitmap data types don't have a `scale` param/instance, because none of them are HiDPI. I wasn't sure if it's a good idea to add instances that only have a constant value - for people just reading the spec it's unnecessary information, and in a visualizer it might be confusing why those values are there.

Reply:
>How would that be possible? icns_style_packbits is compressed data and the byte array case is uncompressed data - they have to be handled differently by code that uses the spec.

I have implemented this kind of stuff once in one of my specs, but I don't remeber which one exactly. I had to use 2 types, one is for compressed and another one for uncompressed. I had to do that because `process` allows no `switch-on`. So for uniformity I had to made an uncompressed one have the same fields as compressed one, switch the wrapped types, and then pass through their content through instances using dirty hacks as `.as<>` to get a bit simpler API. I know that passing through instances duplicates the size, but it is more a bug on compiler side rather than in specs.

Reply:
Turns out that I was wrong and these bitmaps are always compressed - there's no special case for uncompressed bitmaps. The problem was that I originally tested it with a bitmap that (in uncompressed form) is also valid compressed data, and when decompressed it looks very similar to the original, so I didn't notice that the data was misinterpreted.